### PR TITLE
Add Precise Seam placement feature

### LINKED ISSUE #5118: Seam position modifier
### Is there an existing issue for this feature request?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

I have a model with cylindrical and prismatic sections. I would like to have random seams on the cylindrical parts, and aligned/corner seams on the prismatic ones. I thought the obvious solution would be to use a modifier... but surprisingly, modifiers don't support the "seam position" option. Not sure if the reason for this is somehow fundamental, hopefully not.

There are a number of ideas floating around about enabling more control over seams with modifiers (e.g. seam enforcing/seam blocking via modifier instead of only seam painting), which maybe align with this request; in either case, I don't think this specific feature request was clearly articulated before.

### Which printers will be beneficial to this feature?

All

### Describe the solution you'd like

I would like "seam position" to be an available option for a modifier.

### Describe alternatives you've considered

I have considered using seam painting instead, but with global seam mode set to random, even very tightly painted seams will always be rather weakly aligned ("fuzzy").

### Additional context

_No response_

### LINKED ISSUE #11162: High-precision seam placement using modifiers (algorithm)
### Is there an existing issue for this feature request?

- [x] I have searched the existing issues

### Is your feature request related to a problem?

Precisely positioning the seam can be quite a tedious and time-consuming task. Moreover, any change to the model usually requires repeating this work from scratch.  
In CAD systems, however, it’s very easy to create additional bodies that exactly match the desired seam position. You can also use built-in primitives for that purpose.  

<p align="center">
  Image 1
  Image 2
</p>

[Test Objects 3MF.zip](https://github.com/user-attachments/files/23162512/Test.Objects.3MF.zip)

### Which printers will be beneficial to this feature?

All

### Describe the solution you'd like

At the moment, Orca Slicer allows adding arbitrary bodies as modifiers, but it’s not possible to explicitly control seam placement using them.

**Proposed solution**
I suggest a seam-positioning algorithm that’s very easy to implement:

1. Add a property for modifiers: **“Easy seam”**.
2. During slicing, for each layer, obtain the **outer perimeter polygons** of both the main object and the modifiers.
3. Check each object polygon for intersections with the modifier polygons — using **ClipperLib’s `ctIntersection`** operation (via `Clipper::Execute()`), which returns the intersection **paths**. Stop at the first intersection found.
4. If an intersection exists, override all existing seam-placement mechanisms.
5. Take the first intersection path (ignore the rest, if any).
6. Find the points of this path that belong to the object’s polygon — the midpoint of that segment becomes the seam point (see the function description below).

**That’s it.**

*Later we can add more options if needed.*

Image

### Describe alternatives you've considered

_No response_

### 🧮 Function description: Finding the seam point from the intersection path

The goal is to find the **segment of the intersection path** that belongs to the **object polygon** (not the modifier polygon), thus defining the segment whose midpoint becomes the target seam point.

I’m not sure in which order `ctIntersection` returns the points along the intersection path. If the first point is always the intersection of the two original polygons, that would greatly simplify the task.

In the general case, the function could work as follows:

1. Sequentially iterate through the points on the intersection path and look for one that matches a point from the object polygon.  
2. If such a point is found, continue searching to find the **last matching point**. If the first found one was the first point of the path — search the same way in the opposite direction.  
3. For the found point(s), inspect the two adjacent points to determine whether their connecting segments belong to the object polygon. This gives two points that define the boundaries of the target segment.  
4. If no matching points are found, it means the modifier polygon intersected a single edge of the object polygon without including any of its vertices. In that case, check each segment of the intersection path to see if it lies within an edge of the object polygon. The overlapping segment is the one we need.  

Finally, compute the **midpoint of the identified segment** — that’s the desired seam point.

I could implement this as a C++ function that takes two polylines (for the object and the modifier) and returns the resulting seam point. This mainly requires some basic computational geometry.

### PR DESCRIPTION
**Pull request by DImagine**
New modifier type **Precise Seam** that allows users to control exact seam placement on 3D prints using volume-based intersection geometry.

Includes:
- 6 seam modifier subtypes: Center, Left, Right, Enforced, Blocked, Neutral
- Core algorithm: perimeter segment detection via mesh-layer intersection
- SeamPlacer integration: modifiers override seam zone painting on affected perimeters
- GUI: context menus, icons, volume sorting, Change Type support

# Description
**Precise Seam** is a modifier system that gives you precise control over where exactly the seam will be placed on a model during 3D printing. You place a modifier volume on the model, and the seam runs exactly where that volume intersects the model's walls.

The system allows you to either define the exact seam position on the perimeter, or set up allowed and forbidden zones.

For quick results, you can use built-in primitives as modifiers. For complex geometries, you can create a custom auxiliary body once in any 3D modeling software and reuse it permanently. Even for the most complex geometries, placing the seam takes just a couple of clicks. This is especially valuable during **iterative design** — since the modifier is a persistent object, it survives model changes automatically. There is no need to repaint seam zones after each revision. This completely eliminates manual work and provides an unprecedented level of convenience for seam control.

https://github.com/user-attachments/assets/9a8f2d51-b6b9-46c8-8832-fcf2ec02979d

> [!NOTE]
> This modifier body was created by sweeping a triangular cross-section along a curved path on the surface of the mesh. A **circle** is the easiest base profile to use since it doesn't require aligning its orientation. You can also try a regular polygon with enough sides — it has a predictable edge count and works faster, while a circle tends to be unnecessarily over-detailed.

> [!NOTE]
> If you export a model together with auxiliary geometry, change the type of the auxiliary bodies to Modifier before performing any build-plate placement operations on it. Otherwise, automatic placement features (such as "Place on Face") may not work correctly, because conflicts will arise with the faces of the auxiliary bodies.
---

## Modifier Types

Precise Seam includes **6 modifier types**, divided into two groups:

### Strong Modifiers (Exact Positioning)

Define a **specific point** on the perimeter where the seam will be placed. They work strictly: if a Strong modifier intersects a wall — the seam goes exactly there, no alignment strategy can move it.

| Type | Behavior | When to Use |
|------|----------|-------------|
| **Center** | Seam is placed at the **geometric center** of the intersection segment between the modifier and the wall | Universal option. The seam will be in the middle of the intersection area — the most predictable behavior |
| **Left** | Seam is placed at the **beginning** (first point) of the intersection segment | When you need to pin the seam to a specific edge of the intersection |
| **Right** | Seam is placed at the **end** (last point) of the intersection segment | When you need to pin the seam to the opposite edge of the intersection |

### Weak Modifiers (Zone-Based Control)

Do not define a specific point, but **mark zones** on the perimeter, influencing the priority during seam selection by the standard algorithm.

| Type | Behavior | When to Use |
|------|----------|-------------|
| **Enforced** | The intersection zone is marked as **preferred** (green) for the seam | Equivalent to manually painting green seam zones. Use when you want to softly attract the seam to an area without pinning it to a specific point |
| **Blocked** | The intersection zone is marked as **forbidden** (red) for the seam | Equivalent to manually painting red seam zones. Use when you need to remove the seam from a visible surface or critical area |
| **Neutral** | The intersection zone is marked as **neutral** (gray) | Use when you want to reset a zone to the initial neutral state |

## Known Limitations

> [!WARNING]
> The modifier must extend far enough into the object to intersect the outer perimeter’s centerline, which is typically offset inward from the model’s surface by half the outer wall line width.

> [!WARNING]
> Strong modifiers (Center/Left/Right): only one intersection per perimeter is supported, since there can be only one seam. Additional intersections are ignored.
>
> Weak modifiers (Enforced/Blocked/Neutral): multiple intersections are supported, but none of them should pass through the model entirely. A through-body intersection (more precisely, through-printing-perimeter) produces multiple segments, of which only one will be processed.
>
> In both cases, a pop-up warning is shown when unsupported intersections are detected.
>
>If any modifier has a multiply-connected cross-section (e.g. a hollow or torus-like shape), it is skipped and a corresponding notification is shown. This case can still be covered with small workarounds — for example, splitting the modifier into two parts.
>
> Full containment of the perimeter within the modifier is not handled.
> 

---

## Modifier Behavior and Technical Notes

Precise Seam modifiers integrate smoothly into the existing seam placement pipeline. In areas not covered by any modifier, the built-in seam algorithms work unchanged.

The modifiers simply augment (with higher priority) the existing zone-painting mechanism:
- **Strong modifiers**: mark all perimeter points as BLOCKED except for one, which is marked ENFORCED. The built-in algorithms then take over from there.
- **Weak modifiers**: apply automatic zone painting on top of any manual painting, marking points according to their type (Enforced/Blocked/Neutral).

---
## Hierarchy and Priorities

When multiple modifiers are present, their priority is determined by position in the object tree — modifiers higher in the tree have higher priority. Strong modifiers are always positioned above Weak ones in the tree — this is built-in behavior, because Strong always takes precedence over Weak. When a Strong modifier applies, all modifiers below it in the list are ignored for that perimeter.

If multiple Weak modifiers overlap in the same zone, the ones higher in the list take precedence, and the zone is painted in the corresponding "colors" according to this order.

Modifier order can be rearranged via **drag & drop** in the object tree.

---

### Adding a Precise Seam Modifier

Adding a Precise Seam modifier follows the standard modifier workflow: right-click on an object — either in the object tree or on a selected object in the scene — and select **"Add precise seam"** from the context menu, then choose a specific 3D shape for the modifier.

You can also convert any existing 3D object (including other modifiers) in the scene into a Precise Seam modifier. The only requirement is that only objects within a group containing at least one other body can be converted into modifiers — a standalone object outside of any group cannot be converted. Groups and hierarchy are visible in the object tree.

To convert, find the object you want to convert in the object tree and make sure it is within a group. Select it and verify visually in the scene that it is the correct object. Then right-click it (in the scene or in the tree), select **"Change type"** from the context menu, and choose **"Precise Seam"** from the dropdown list.

https://github.com/user-attachments/assets/b29c87ea-adce-445c-b5ba-145f90e792c2

### Changing the Modifier Type

Once you have a Precise Seam modifier, you can change its type to any of the six available variants. Select the modifier in the tree or in the scene, right-click, and choose **"Precise Seam Type"** from the context menu — it will show a list of all Precise Seam modifier types. Select the one you need.

Context Menu

### Visual Distinction of Types

Each Precise Seam type has its own color in the 3D view (all at 60% transparency):

| Type | Color | Description |
|------|-------|-------------|
| Center | 🟠 Orange | Warm orange — primary Strong type |
| Left | 🟡 Gold | Bright golden |
| Right | 🟠 Dark Orange | Deep orange |
| Enforced | 🟢 Green | Green — "allowed" zone |
| Blocked | 🔴 Red | Red — "forbidden" zone |
| Neutral | ⚪ Gray | Gray — neutral zone |

Strong modifiers are visually unified with warm orange tones. Weak modifiers use intuitive colors: green = allowed, red = forbidden, gray = neutral.

###  YouTube Full Demo Video

Demonstrates the behavior of all six modifier subtypes and their application hierarchy based on their position in the object tree.

[Precise Seam Demo](https://www.youtube.com/watch?v=F-In1N7AquQ)

## Test Files

[Test Bell Object.zip](https://github.com/user-attachments/files/26289912/Test.Bell.Object.zip)  This test file was created by drawing a spline curve on the surface of the printed object, then sweeping a triangular profile along it (Swept Boss/Base in SolidWorks). This way the seam follows the desired trajectory precisely. A circular profile can also be used as the sweep cross-section — this avoids orientation issues.

[Big Test.zip](https://github.com/user-attachments/files/26383749/Big.Test.zip)  Another file demonstrating all modifier types.

> [!NOTE]
> These test files have only been verified on Windows. Testing on Linux (AppImage/Flatpak) and macOS would be greatly appreciated — please share your results in the comments.

## Debug Instruments
A debug branch that dumps geometry data to files:
https://github.com/DImagine/OrcaSlicer/tree/feature-precise-seam-debug

Python scripts for data visualization:
[Precise Seam Python Visual Debug.zip](https://github.com/user-attachments/files/26330531/Precise.Seam.Python.Visual.Debug.zip)

## Feature Development Progress Log

Progress data was logged at each stage of feature development. This information is included for reference, as it may be helpful for understanding the algorithm's internals.
[Precise Seam Progress Log.zip](https://github.com/user-attachments/files/26289946/Precise.Seam.Progress.Log.zip)

## Related Issue
Closes #11162
Closes #5118

### May also help
These issues describe different approaches to the same underlying problem — explicit seam placement control. This PR provides a different mechanism (modifier objects), which may be useful for the reporters:

- #10064
- #8884
- #7799
- #11134
- #8427
- #5621

## Changelog
- 2026-03-27: Added handling for multiple intersection edge case
- 2026-03-28: Added pop-up warning message if multiple intersections are detected
- 2026-03-28: Improved detection and handling of multiple intersections, refined user notifications for problematic modifiers
- 2026-03-29: Some minor updates and improvements
- 2026-03-31: Refactor PreciseSeam based on AI code review findings
- 2026-04-10: Fix raft offset in Precise Seam cache indexing, add legacy modifier flag in 3MF export. Based on Codex code review.
- 2026-04-12: Fix 3MF round-trip, refine_at_vertex direction, PrintApply cross-family asserts, multi-select type dialog, GUI config paths and SLA clipboard for Precise Seam. Based on Codex code review.
- 2026-04-17: Fixes from Claude Opus 4.7 code review; false positives in accompanying file; rebased onto latest `main`.
[Code Review - Known False Positives.md](https://github.com/user-attachments/files/26803538/Code.Review.-.Known.False.Positives.md)
- 2026-04-18: Restored Precise Seam in the Change Type context-menu flow after upstream `d7dbe6a0ed` replaced the modal dialog with a submenu. Added "Precise Seam" entry to the new Change Type submenu (with subtype preservation on already-PS volumes) and reworked the "Precise Seam Type" subtype submenu to support multi-select with correct PS group-reorder.
- 2026-04-20: Isolated dead-code cleanup into #13281; restored ObjectList::change_part_type() and its refs, narrowed the SVG/text guard in set_volume_type() to PS-only via is_precise_seam().
- 2026-04-25: Rebased onto upstream/main (incl. merged #13281); kept upstream's `#if 0` change_part_type() verbatim and extended the SUPPORT_* guard in set_volume_type() with `|| is_precise_seam(new_type)`. Squashed 28 commits into one.
- 2026-04-27: Rebased on top of #13120. Extended its SVG/text UI-guard to also cover
Precise Seam types. Minor wording tweak in one slicing warning.

### CODE REVIEW

**Comment by DImagine** [comment]
@igiannakas @RF47 @vovodroid @valerii-bokhan @EaglePPP @MxBrnr @discip I'd appreciate your feedback on this feature if you get a chance.

**Comment by Ataraxia-Mechanica** [comment]
This would be an excellent feature for me, as I usually use blender and generating a separate seam model from existing vertices is trivial.

**Comment by vovodroid** [comment]
> I'd appreciate your feedback on this feature if you get a chance.

Indeed, good idea! Upvote.

**Comment by igiannakas** [comment]
Love it! Quite interesting and innovative! 

However to account for user error we do need to consider how we handle fall back scenarios where the modifier volume intersects incorrectly in more than one places.

**Comment by Ataraxia-Mechanica** [comment]
I made a simple blender geometry node modifier that allows extracting edges on the model as a separate seam model, just right click the edges you want, select mark freestyle edge:
image
image
image

generated model won't be manifold if the selected edge has forks in it unfortunately, but in most cases a fork isn't needed.

[seam geometry node.zip](https://github.com/user-attachments/files/26304358/seam.geometry.node.zip)

**Comment by DImagine** [comment]
> to account for user error

I will try to figure out something...
Can we agree that if there are multiple intersections, one random region is always selected? There will be no errors, and the behavior is predictable. This covers 99.9% of use cases. The current algorithm is very fast, but rewriting it to correctly handle all intersections would significantly slow it down — not to mention requiring substantial development time.

**Comment by Ataraxia-Mechanica** [comment]
> > to account for user error
> 
> I will try to figure out something... Can we agree that if there are multiple intersections, one random region is always selected? There will be no errors, and the behavior is predictable. This covers 99.9% of use cases. The current algorithm is very fast, but rewriting it to correctly handle all intersections would significantly slow it down — not to mention requiring substantial development time.

selecting the one that's closer to seams on adjacent layers would be ideal but one random region would be fine in my opinion

**Comment by DImagine** [comment]
I improved the algorithm: multiple intersections are now handled gracefully by selecting one. Still, for fully correct results the user must be aware that this limitation remains. The initial source of this limitation is that there can be only one seam per perimeter.

**Comment by Xipit** [comment]
Why not add an error or warning message when multiple intersections are present?
Similar to when objects cross the bed boundary.

**Comment by DImagine** [comment]
> Why not add an error or warning message when multiple intersections are present?

I think this is a good option. If the community and maintainers support this approach, it can be implemented.

**Comment by MxBrnr** [comment]
Interesting idea, but to be honest, I will likely never use this feature.

**Comment by DImagine** [comment]
Added a pop-up warning message when multiple intersections between the object and modifier are detected. Works like a charm. Thanks @Xipit for the idea!


Warning

**Comment by DImagine** [comment]
Pushed update: improved detection and handling of multiple intersections, refined user notifications for problematic modifiers.

**Comment by BigBOSSJake** [comment]
Sounds like a good thing for seams to be more controllable as they can be quite ugly.

**Comment by DImagine** [comment]
Pushed update: some minor updates and improvements.

**Comment by DImagine** [comment]
@Ataraxia-Mechanica @dmitry-sorkin Could you try using this feature on some real-life models? I think in skilled hands like yours, it can save a huge amount of time when precise seam placement is needed. Just note that the modifier must intersect the actual printed perimeter, which is offset inward by half the layer width from the wall — so the modifier needs to penetrate deep enough. Binaries (including portable) are available in the release on [this branch](https://github.com/DImagine/OrcaSlicer/tree/feature-precise-seam-pr) or in this PR. Try not to push this feature to its very limits at the very first tests :)

**Comment by Ataraxia-Mechanica** [comment]
> @Ataraxia-Mechanica @dmitry-sorkin Could you try using this feature on some real-life models? I think in skilled hands like yours, it can save a huge amount of time when precise seam placement is needed. Just note that the modifier must intersect the actual printed perimeter, which is offset inward by half the layer width from the wall — so the modifier needs to penetrate deep enough. Binaries (including portable) are available in the release on [this branch](https://github.com/DImagine/OrcaSlicer/tree/feature-precise-seam-pr) or in this PR. Try not to push this feature to its very limits at the very first tests :)

image
image
image
image


Enforced type looks fine when pushing a bit of limits to me, all seams snapped to edges when there's a intersection point in the wall loop, and painting alongside it works fine too. Some seams doesn't show up as white dots but they are in the right place when dragging the timeline at the bottom, that was a regression caused by something else and not this PR.

**Comment by DImagine** [comment]
> a regression caused by something else

Please make sure the modifier runs top-to-bottom in a single pass (like the green line). I suspect your model may actually match the red-line case — that is, the modifier intersects the same perimeter twice. If that's the case, the seam chain could indeed break, because only one of the two intersection points would be selected.

<div align="center">
  image
</div>

I may be wrong, but could you double-check? Also — do you get any pop-up warnings in the bottom-right corner during slicing?

The only time I've seen seam points fail to form a clean chain was when **Scarf Joint Seam (beta)** was enabled in the settings — but in that case the same issue occurred even *without* my modifier, so it clearly wasn't related to it.

Would it be appropriate to ask you to share the model as a **.3MF file** for testing purposes? Including both the main body and the modifier.

**Comment by DImagine** [comment]
@Ataraxia-Mechanica You really pushed the feature to its limits :) Try making a short modifier that would match the white seam line in your drawing. Let's start with something simple :)
<div align="center">
  image
</div>

**Comment by Ataraxia-Mechanica** [comment]
> > a regression caused by something else
> 
> Please make sure the modifier runs top-to-bottom in a single pass (like the green line). I suspect your model may actually match the red-line case — that is, the modifier intersects the same perimeter twice. If that's the case, the seam chain could indeed break, because only one of the two intersection points would be selected.
> image
> 
> I may be wrong, but could you double-check? Also — do you get any pop-up warnings in the bottom-right corner during slicing?
> 
> The only time I've seen seam points fail to form a clean chain was when **Scarf Joint Seam (beta)** was enabled in the settings — but in that case the same issue occurred even _without_ my modifier, so it clearly wasn't related to it.
> 
> Would it be appropriate to ask you to share the model as a **.3MF file** for testing purposes? Including both the main body and the modifier.

Yes the modifier does intersect more than once in some perimeters and the warning does show up, but looks like it's fine with "enforce" mode as they do line up, similar to what would happen when multiple areas are painted manually, checked about ~20 layers where white dot doesn't show up and they all aligned at the same area. 
There are areas where seam would drift away a bit though, they are probably caused by unoptimal seam parameters, as they would do this with seam paints too, and would probably only be solved by using strong modifier mode or tweaking parameters like #12087.
image
[Seamtest.zip](https://github.com/user-attachments/files/26333497/Seamtest.zip)

**Comment by DImagine** [comment]
> Seamtest.zip

Unfortunately, it turns out that SolidWorks can't handle .3MF files properly… and I don't really know my way around Blender either. (I belive you use Blender)

My thinking is this: from the existing long modifier, we just need to isolate the region shown in the drawing — so that it only passes top-to-bottom in a single pass. The cross-section diameter could be something like 1 mm, just enough to cleanly intersect the perimeter. Set the modifier to CENTER, and everything should work fine.

Screenshot_1

If it's not too much trouble, could you upload the model (with the modifier) in Blender format? I'll try to figure it out when I get a chance and make the perfect seam.

**Comment by DImagine** [comment]
> Seamtest.zip

@Ataraxia-Mechanica Here is the body you tried to apply as seam placer. It's to much, my friend :)

image

Cut out a single segment from this long shape — one that runs strictly top-to-bottom along its entire length, without creating multiple intersections with the same perimeter. That way the seam position will be unambiguous, the pop-up warnings will stop appearing, and you'll get a clean, consistent seam right where you want it.

**Comment by DImagine** [comment]
> that was a regression caused by something else and not this PR.

I sliced the model with your settings — even without my modifier — and I also got those strange seam gaps. So the modifier isn't the cause.

I then sliced your model with my own printer settings and am attaching the result. I intentionally set the seam to Random so I could clearly see how the modifier performed. Despite operating in a non-standard mode due to multiple perimeter intersections, it still managed to place the seams precisely along that sophisticated trajectory on every perimeter it influenced. No seam gaps were observed.

<div align="center">
  
</div>


The algorithm isn't perfect — it requires you to carefully plan the seam path first, and only then build the modifier to position it. But it works.

*BTW, you are really cool at extracting such sophisticated trajectories! Do you do it with some scripts?*

[Seamtest-random.zip](https://github.com/user-attachments/files/26335169/Seamtest-random.zip)

**Comment by Ataraxia-Mechanica** [comment]
> > that was a regression caused by something else and not this PR.
> 
> I sliced the model with your settings — even without my modifier — and I also got those strange seam gaps. So the modifier isn't the cause.

Time for me to check what setting is causing this and make a bug report then, thanks for helping me figure this out.

I used the geometry node modifier thingy I posted all the way up, it's basically blender's visual programming language, here's the blender project of the model, set the subdivision modifier to 1 level to avoid lagspikes when editing:
[mouse-mittens-v1.zip](https://github.com/user-attachments/files/26338462/mouse-mittens-v1.zip)

**Comment by DImagine** [comment]
> Here is the body you tried to apply as seam placer. It's to much, my friend :)

@Ataraxia-Mechanica I cut out a bounded volume from your sophisticated modifier and applied it separately. The seam landed exactly where it needed to.

ScreenshotXX_2

ScreenshotZ_2

**Comment by DImagine** [comment]
I completed a code review with AI assistance and will gradually implement several changes. 

**Pushed update:** Cache modifier slices to avoid redundant re-slicing per perimeter

**Comment by DImagine** [comment]
@RF47 I believe this PR is ready for review.

1. **Code quality.** The code has been through an AI-assisted review — all findings have been addressed.

2. **Thoroughly tested.** Extensive manual testing across various models and modifier configurations. With correct modifier placement, seams always land exactly where expected.

3. **Robust error handling.** Invalid modifier geometry (through-body intersections, multiply-connected regions, etc.) never causes crashes — the algorithm handles it gracefully and shows user-facing warnings.

4. **Clean integration.** Existing seam placement logic is not touched — Precise Seam only activates when its modifiers are present.

Would appreciate your review when you get a chance.

**Comment by RF47** [comment]
> @RF47 I believe this PR is ready for review.
> 
> 1. **Code quality.** The code has been through an AI-assisted review — all findings have been addressed.
> 2. **Thoroughly tested.** Extensive manual testing across various models and modifier configurations. With correct modifier placement, seams always land exactly where expected.
> 3. **Robust error handling.** Invalid modifier geometry (through-body intersections, multiply-connected regions, etc.) never causes crashes — the algorithm handles it gracefully and shows user-facing warnings.
> 4. **Clean integration.** Existing seam placement logic is not touched — Precise Seam only activates when its modifiers are present.
> 
> Would appreciate your review when you get a chance.

I think you're overestimating me—there are people here who know a lot more than I do, especially when it comes to reviewing a PR of this magnitude—but I'm happy to help in any way I can, and I'll take a look at it. Just keep in mind that I don't have the authority to merge something this complex.

File: src/slic3r/GUI/GUI_ObjectList.hpp [hunk 1]
```diff
@@ -85,6 +85,16 @@ struct MeshErrorsInfo
...
 {
 public:
+
+    struct VolumeHierarchyEntry
+    {
+        // For current UI hierarchy snapshot.
+        const ModelObject* object { nullptr };       // Owning object node.
+        const ModelVolume* volume { nullptr };       // Current volume entry.
+        const ModelVolume* parent_volume { nullptr };// Direct parent volume, null when attached to object.
+        int                order_within_parent { -1 }; // Position among siblings matching UI order.
```
[the remark this answers is not in the document]
Comment (DImagine) [reply src/slic3r/GUI/GUI_ObjectList.hpp]: Agreed that the comment was inaccurate. However, Copilot's suggested wording "may differ from UI sibling order" is misleading — `volume_idx` *is* monotonic among siblings and mirrors UI order, it's just not a dense 0..N-1 index. Updated the comment to:
```cpp
int order_within_parent { -1 }; // Volume index in ModelObject::volumes; monotonic within siblings, mirrors UI order.
```

File: src/libslic3r/GCode/PreciseSeam.hpp [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+// Pre-sliced modifier cache: ModelVolume pointer → per-layer Polygons.
+// Built once in SeamPlacer::init(), then passed read-only into per-perimeter functions.
+using ModifierSlicesCache = std::unordered_map<const ModelVolume*, std::vector<Polygons>>;
+
+// Warning flags set during Precise Seam processing (thread-safe)
+struct PreciseSeamWarnings {
+    std::atomic<bool> multiple_intersections{false};  // multiple or through-body intersections detected
+    std::atomic<bool> multiply_connected{false};      // modifier has holes (multiply-connected cross-section)
+};
```
[the remark this answers is not in the document]
Comment (DImagine) [reply src/libslic3r/GCode/PreciseSeam.hpp:61]: Valid. Added `#include <atomic>` to make the header self-contained.

File: src/slic3r/GUI/GUI_ObjectList.cpp [hunk 3]
```diff
@@ -1870,6 +1930,24 @@ bool ObjectList::can_drop(const wxDataViewItem& item, int& src_obj_id, int& src_
...
+
+        // Special handling for Precise Seam modifiers: allow drag&drop within same group (strong↔strong or weak↔weak)
+        const int obj_idx = m_dragged_data.obj_idx();
+        const int dragged_vol_idx = m_dragged_data.sub_obj_idx();
+        const int target_vol_idx = m_objects_model->GetVolumeIdByItem(item);
+
+        if (obj_idx >= 0 && dragged_vol_idx >= 0 && target_vol_idx >= 0) {
+            ModelVolume* dragged_vol = (*m_objects)[obj_idx]->volumes[dragged_vol_idx];
+            ModelVolume* target_vol  = (*m_objects)[obj_idx]->volumes[target_vol_idx];
+
```
[the remark this answers is not in the document]
Comment (DImagine) [reply src/slic3r/GUI/GUI_ObjectList.cpp]: Valid as defensive programming. The surrounding code in this method uses the same unchecked pattern (line 1956), so the risk is systemic rather than specific to our code. Nevertheless, added explicit bounds checks for `obj_idx` against `m_objects->size()` and both volume indices against `volumes.size()`. On out-of-range, returns `false` (drop disallowed).

**Comment by DImagine** [comment]
@RF47 Thanks for running the Copilot review! It came back with just 3 minor remarks across the entire 2.5K-line PR — one was a comment clarification, one a missing include, and one a defensive bounds check. All three are addressed and pushed.

What would be the best next step? Should I just wait, or is there anything you could do to help get eyes on this from someone with merge authority?

**Comment by valerii-bokhan** [comment]
Hey, sorry, I just found some time to test it. I'm grabbing the code now.

**Comment by RF47** [comment]
> @RF47 Thanks for running the Copilot review! It came back with just 3 minor remarks across the entire 2.5K-line PR — one was a comment clarification, one a missing include, and one a defensive bounds check. All three are addressed and pushed.
> 
> What would be the best next step? Should I just wait, or is there anything you could do to help get eyes on this from someone with merge authority?

You'll need to be patient; keep in mind that there are already more than 300 open pull requests; ideally, if possible, we should test this on Linux using AppImage and Flatpak, and on Mac.

**Comment by DImagine** [comment]
> You'll need to be patient; keep in mind that there are already more than 300 open pull requests; ideally, if possible, we should test this on Linux using AppImage and Flatpak, and on Mac.

Thanks for the guidance! This is my first PR and I'm still learning how things work in the community, so I appreciate the patience.

I noticed some PRs have a "Community testers wanted" label — would it make sense to add that here? I only have a Windows setup, so I can't test Linux/Mac builds myself, but I've attached test project files in the PR description that should make it easy for anyone to verify the feature.

**Comment by RF47** [comment]
> "Community testers wanted" label

done

**Comment by valerii-bokhan** [comment]
> Hey, sorry, I just found some time to test it. I'm grabbing the code now.

Well, I need to find a test case for this feature.

I mostly print miniatures that are usually very complex organic geometry with many fine details. It's a bit hard to say how this feature can utilized in this case.

**Comment by RF47** [comment]
I ran a simple test on Linux using Flatpak and AppImage (a virtual machine running Debian)
It didn't crash, which is good
image
image

[test seamplacer.zip](https://github.com/user-attachments/files/26390610/test.seamplacer.zip)

**Comment by DImagine** [comment]
@RF47 Thanks for the great test! I modified it slightly to demonstrate how smoothly Precise Seam integrates with the existing code.

Under the hood, when a strong modifier is used, one point on the perimeter is marked as ENFORCED and all others as BLOCKED. This lets all the built-in seam placement algorithms work as if the user had manually painted this point with the brush tool — we only intervene at the "zone painting" stage.

I moved your modifier slightly so that it doesn't intersect the body at the concave areas, then ran two slices — one with Random seam and one with Aligned. In the Random case, the seam is aligned only where the modifier intersects the object. But in the Aligned case, thanks to the built-in alignment mechanisms, the seam stays aligned even in the concave areas.

<table>
  <tr>
    <td>Seam3</td>
    <td>Seam1</td>
  </tr>
  <tr>
    <td>Seam4</td>
    <td>Seam2</td>
  </tr>
</table>

[test seamplacer shifted.zip](https://github.com/user-attachments/files/26395815/test.seamplacer.shifted.zip)

**Comment by DImagine** [comment]
@valerii-bokhan I'd really appreciate it if you could just play around with this feature — see how it works, whether the results feel intuitive, and whether the workflow is user-friendly. There's no need to look for specific edge cases. You can create any modifier in a 3D modeling app or simply use built-in primitives. It would also be great if someone could test it on platforms other than Windows, especially macOS.

**Comment by DImagine** [comment]
Pushed update: fix raft offset in Precise Seam cache indexing, add legacy modifier flag in 3MF export. Based on Codex code review.

**Comment by DImagine** [comment]
**Code Quality & Testing**

The code has been reviewed through multiple iterations of automated code review using ChatGPT 5.4 and Claude Opus 4.6. These models dig remarkably deep, catching edge cases and subtle issues — I'm confident the current code is near production-perfect.

**Personal Experience**

I use this feature extensively in my own iterative design workflow. You place a helper body once, and seam positioning takes just a couple of clicks. For anyone familiar with 3D modeling software, creating such bodies — even complex ones — is trivial: you draw a line on the surface in a few clicks, then sweep a circular profile along it. The seam placement accuracy is absolute. Beyond saving time, there's a key benefit: you never forget to place the seam correctly on the next iteration. The figure below shows a slice where seam positioning was set to **random**, yet this particular seam was placed precisely using Precise Seam.

bake

**Comment by DImagine** [comment]
**Pushed update:**

Following the recent release of Claude Opus 4.7 by Anthropic, I ran an in-depth review of this PR with the new model. Several fixes identified by that review have been applied. Across multiple automated review sessions a set of false positives has accumulated as well — documented in the accompanying file for reference. The branch has also been rebased onto the latest `main` to stay current.

[Code Review - Known False Positives.md](https://github.com/user-attachments/files/26803538/Code.Review.-.Known.False.Positives.md)

**Comment by DImagine** [comment]
@RF47, thanks for linking #5118! I also found another one — #10064 — which requests seam placement modifiers to avoid manual re-placement of seams across design iterations.

**Comment by DImagine** [comment]
**Pushed:** restored Precise Seam in the Change Type context-menu flow after the upstream rework (`d7dbe6a0ed`) replaced the modal dialog with a submenu. Added a "Precise Seam" entry to the new Change Type submenu (with subtype preservation on already-PS volumes) and reworked the "Precise Seam Type" subtype submenu to support multi-select with correct PS group-reorder.

**Comment by DImagine** [comment]
**Pushed update: scoped the diff down to Precise Seam only.**

- Restored `ObjectList::change_part_type()` and its refs — dead-code cleanup is now isolated into #13281
- Narrowed the SVG/text guard in `set_volume_type()` to PS-only (via `is_precise_seam()`); the `SUPPORT_*` coverage lives in #13281
- Diff stats: ~200 fewer deletions

**Comment by RF47** [comment]
@DImagine It would be good if you could resolve the conflicts so it can be recompiled.

**Comment by DImagine** [comment]
> @DImagine It would be good if you could resolve the conflicts so it can be recompiled.

@RF47 Done — rebased on the latest upstream/main and resolved all conflicts. Builds cleanly locally on Windows. Ready for another look whenever you have time.

**Comment by DImagine** [comment]
**Pushed update:** Rebased on top of #13120. Extended its SVG/text UI-guard to also cover
Precise Seam types. Minor wording tweak in one slicing warning.
